cirq-core/cirq/devices/noise_properties.py

```diff
@@ -234,37 +42,69 @@ def __init__(self, noise_properties: NoiseProperties) -> None:
+    def virtual_predicate(self, op: 'cirq.Operation') -> bool:
+        """Returns True if an operation is virtual.
 
-        if any(
-            [protocols.is_measurement(op.gate) for op in moment.operations]
-        ):  # Add readout error before measurement gate
-            p00 = self._noise_properties.p00
-            p11 = self._noise_properties.p11
-            measurement_qubits = [
-                list(op.qubits)[0] for op in moment.operations if protocols.is_measurement(op.gate)
+        Device-specific subclasses should implement this method to mark any
+        operations which their device handles outside the quantum hardware.
+        """
+        return False
+
+    def noisy_moments(
+        self, moments: Iterable['cirq.Moment'], system_qubits: Sequence['cirq.Qid']
+    ) -> Sequence['cirq.OP_TREE']:
+        # Hack to bypass import order constraints
+        from cirq import circuits
+
+        # Split multi-qubit measurements into single-qubit measurements.
+        # These will be recombined after noise is applied.
+        split_measure_moments = []
+        multi_measurements = {}
+        for moment in moments:
+            split_measure_ops = []
+            for op in moment:
+                if not protocols.is_measurement(op):
+                    split_measure_ops.append(op)
+                    continue
+                m_key = protocols.measurement_key_obj(op)
+                multi_measurements[m_key] = op
+                for q in op.qubits:
+                    split_measure_ops.append(ops.measure(q, key=m_key))
+            split_measure_moments.append(ops.Moment(split_measure_ops))
+
+        new_moments = []
+        for moment in split_measure_moments:
+            virtual_ops = {op for op in moment if self.virtual_predicate(op)}
+            physical_ops = [
+                op.with_tags(PHYSICAL_GATE_TAG) for op in moment if op not in virtual_ops
             ]
-            if p00 is not None or p11 is not None:
-                _apply_readout_noise(p00, p11, moments, measurement_qubits)
-            moments.append(moment)
-        else:
-            moments.append(moment)
-        if self._noise_properties.pauli_error is not None:  # Add depolarization error#
-            duration = max([get_duration_ns(op.gate) for op in moment.operations])
-            pauli_error = self._noise_properties.pauli_error_from_depolarization(duration)
-            _apply_depol_noise(pauli_error, moments, system_qubits)
-
-        if self._noise_properties.t1_ns is not None:  # Add amplitude damping noise
-            duration = max([get_duration_ns(op.gate) for op in moment.operations])
-            _apply_amplitude_damp_noise(
-                duration, self._noise_properties.t1_ns, moments, system_qubits
-            )
-        return moments
+            # Only subclasses will trigger this case.
+            if virtual_ops:
+                new_moments.append(ops.Moment(virtual_ops))  # coverage: ignore
```

Comment:
Is this because we expect subclasses to override what's virtual and what is not ?

Reply:
Yup, both on the comment and the `coverage: ignore`.